tests/integration/utils/db_faker.py

```diff
@@ -0,0 +1,244 @@
+        defaults = {
+            'cycle': task.cycle,
+            'name': task.name,
+            'try_num': 1,
+        }
+        task_jobs = task.jobs or []
+        if task.jobs is None:
+            for submit_num in range(1, task.submit_num or 1):
+                task_jobs.append(Job('failed'))
+            if _task_has_job(task.status):
+                task_jobs.append(Job(task.status))
+        for job_def in task_jobs:
+            defaults['submit_num'] = submit_num
+            job_data = {}
+            if job_def.status:
+                job_data = JOB_STATUS_MAP[job_def.status]
+            jobs.append(
+                Job(**{
+                    'submit_num': submit_num,
+                    **job_def._asdict(),
+                    **job_data,
+                    **defaults,
+                })
+            )
+            submit_num += 1
+
+    return jobs
+
+
+def _mock_submit_num(tasks):
+    """Determine the submit numbers for the provided tasks."""
+    for task in list(tasks):
+        if task.submit_num is None and task.jobs:
+            tasks.remove(task)
+            tasks.append(
+                Task(**{
+                    **task._asdict(),
+                    'submit_num': len(task.jobs)
+                })
+            )
+        elif task.submit_num is None and _task_has_job(task.status):
+            tasks.remove(task)
+            tasks.append(
+                Task(**{
+                    **task._asdict(),
+                    'submit_num': 1
+                })
+            )
+    return tasks
+
+
+def _insert(conn, table, items):
+    """Insert `items` into `table` using `conn`."""
+    fields = [
+        item[0]
+        for item in CylcSuiteDAO.TABLES_ATTRS[table]
+    ]
+    conn.executemany(
+        f'''
+            INSERT INTO
+                {table} ({','.join(fields)})
+            VALUES
+                ({','.join(['?'] * len(fields))})
+        ''',
+        [
+            tuple(
+                getattr(item, field)
+                for field in fields
+            )
+            for item in items
+        ]
+    )
+
+
+def fake_db(tasks, path='db'):
+    f"""Fake a Cylc run database inserting the provided tasks.
+
+    Args:
+        tasks (list):
+            List of Task instances.
+
+            If a task provides a list of jobs they will also be inserted into
+            the database.
+
+            Otherwise we will attempt to guess what jobs should be associated
+            with a task.
+        path (str):
+            The path to create the database in.
+
+    See the comments at the top of {__file__} for information about which
+    fields must be set and which can be guessed for both tasks and jobs.
+
+    """
+    tasks = _mock_submit_num(tasks)
+    jobs = _mock_jobs(tasks)
+    dao = CylcSuiteDAO(path)
+    try:
+        _insert(dao.conn, 'task_pool', tasks)
+        _insert(dao.conn, 'task_states', tasks)
+        _insert(dao.conn, 'task_jobs', jobs)
```

Comment:
Anything else we should mock.

SoD has given us the nice curveball that is `absolute_outputs`.

Reply:
Why is that a curveball?

Reply:
It's task info which has been selected based on graphing criterion but transposed into outputs which is just a bit different to the old approach.